# Check Fields for CDLE

In [1]:
import pandas as pd

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', None)

In [2]:
fin = open("list")
files=fin.readlines()
files

['Ces.csv\n',
 'Income.csv\n',
 'Industry.csv\n',
 'Labforce.csv\n',
 'OESWage.csv\n',
 'lngproj.csv\n',
 'shproj.csv\n']

## Read New Data

In [3]:
path = "/home/joe/bic_etl/cdle/cdle_LaborAndEmployment/data_transformed/"
dfs={}
for file in files:
    file=file.strip()
    dfs[file] = pd.read_csv(f"{path}0824/{file}")
    
    print(file,dfs[file].shape)

/tmp/ipykernel_1256/4147802200.py:5: DtypeWarning: Columns (2,5,6,7,9,10,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29) have mixed types. Specify dtype option on import or set low_memory=False.
  dfs[file] = pd.read_csv(f"{path}0824/{file}")


Ces.csv (155584, 1)
Income.csv (9024, 1)


/tmp/ipykernel_1256/4147802200.py:5: DtypeWarning: Columns (2,4,6,7,8,9,11,16,17,18,19,20,21,22,23,24,25,26,27,28) have mixed types. Specify dtype option on import or set low_memory=False.
  dfs[file] = pd.read_csv(f"{path}0824/{file}")


Industry.csv (1394832, 1)
Labforce.csv (35920, 1)


/tmp/ipykernel_1256/4147802200.py:5: DtypeWarning: Columns (2,4,6,7,8,10,11,12,13,14,15,16,17) have mixed types. Specify dtype option on import or set low_memory=False.
  dfs[file] = pd.read_csv(f"{path}0824/{file}")
/tmp/ipykernel_1256/4147802200.py:5: DtypeWarning: Columns (2,4,6,7,8,10,11,12,14,16,18,20,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39) have mixed types. Specify dtype option on import or set low_memory=False.
  dfs[file] = pd.read_csv(f"{path}0824/{file}")


OESWage.csv (329412, 1)


/tmp/ipykernel_1256/4147802200.py:5: DtypeWarning: Columns (2,3,5,7,9,10,11,13,14,16,17,18,19,20,21,22,23,24,25,26,27,28) have mixed types. Specify dtype option on import or set low_memory=False.
  dfs[file] = pd.read_csv(f"{path}0824/{file}")


lngproj.csv (96799, 1)
shproj.csv (91200, 1)


/tmp/ipykernel_1256/4147802200.py:5: DtypeWarning: Columns (2,3,5,7,9,10,11,13,14,16,17,18,19,20,21,22,23,24,25,26,27,28) have mixed types. Specify dtype option on import or set low_memory=False.
  dfs[file] = pd.read_csv(f"{path}0824/{file}")


In [4]:
path = "/home/joe/bic_etl/cdle/cdle_LaborAndEmployment/data_transformed/"
dfs3={}
dfs4={}
for file in files:
    file=file.strip()
    income4 = pd.read_csv(f"{path}0824/{file}")
    income3 = pd.read_csv(f"{path}0824/{file}")
    dfs3[file]=income3
    dfs4[file]=income4
    
    print(file,income3.shape,income4.shape,income3.shape[0]-income4.shape[0])

Ces.csv (153215, 30) (154206, 30) -991
Income.csv (9022, 19) (9022, 19) 0
Industry.csv (1361604, 29) (1372736, 29) -11132
Labforce.csv (35366, 18) (35598, 18) -232


/home/joe/anaconda3/envs/bic/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3072: DtypeWarning: Columns (16) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


OESWage.csv (284676, 40) (284676, 40) 0
lngproj.csv (96797, 29) (96797, 29) 0
shproj.csv (262356, 29) (165772, 29) 96584


## Xrefs

In [4]:
xrefsTitles= {
"2cpa-vbur":"Income-Data-for-Colorado-Counties",
"busm-qa5b":"Occupational-Employment-Statistics",
"cjkq-q9ih ":"Employment-by-Industry-from-Census-of-Employment",
"4e3w-qire":"Employment-and-Unemployment-Estimates",
"pt2g-89wc":"Current-Employment-Statistics",
"gyeb-jc69":"Long-Term Employment Projections in Colorado",
"u2t6-bfhr":"Short-Term Employment Projections in Colorado"}

xrefsFiles = {
"2cpa-vbur":"Income.csv",
"busm-qa5b":"OESWage.csv",
"cjkq-q9ih":"Industry.csv",
"4e3w-qire":"Labforce.csv",
"pt2g-89wc":"Ces.csv",
"gyeb-jc69":"lngproj.csv",
"u2t6-bfhr":"shproj.csv"}

xrefs4x4s = {val:key for key,val in xrefsFiles.items()}

## Read CIM Data

In [17]:
dfs = {}
for file in files:
    file=file.strip()
    w4x4=xrefs4x4s[file]
    wfile=f"https://data.colorado.gov/resource/{w4x4}.csv?$limit=10000000"
    print(file,w4x4,wfile)
    
    df = pd.read_csv(wfile)
    dfs[file]=df

NameError: name 'files' is not defined

In [7]:
for file,df in dfs.items():
    print(file,df.columns)

Ces.csv Index(['stateabbrv', 'statename', 'stfips', 'areatyname', 'areaname', 'area',
       'periodyear', 'periodtype', 'pertypdesc', 'period', 'seriescode',
       'seriesttls', 'seriesdesc', 'adjusted', 'benchmark', 'prelim', 'empces',
       'empprodwrk', 'empfemale', 'hours', 'earnings', 'hourearn',
       'supprecord', 'supphe', 'supppw', 'suppfem', 'hoursallwrkr',
       'earningsallwrkr', 'hourearnallwrkr', 'suppheallwrkr'],
      dtype='object')
Income.csv Index(['stateabbrv', 'statename', 'stfips', 'areatyname', 'areaname',
       'areatype', 'area', 'periodyear', 'periodtype', 'pertypdesc', 'period',
       'inctype', 'incdesc', 'incsource', 'incsrcdesc', 'income', 'incrank',
       'population', 'releasedate'],
      dtype='object')
Industry.csv Index(['stateabbrv', 'statename', 'stfips', 'areaname', 'areatype',
       'areatyname', 'area', 'periodyear', 'periodtype', 'period',
       'pertypdesc', 'indcodty', 'indcode', 'codetitle', 'ownership',
       'ownertitle', 'preli

In [17]:
print(dfs['shproj.csv']['perioddesc'].value_counts())
print(dfs3['shproj.csv']['perioddesc'].value_counts())
print(dfs4['shproj.csv']['perioddesc'].value_counts())

Colorado Short-Term Projections, 2016-2018, June 2016        372870
Colorado Short-Term Projections, 2020-2022, December 2019     86350
Name: perioddesc, dtype: int64
Colorado Short-Term Projections, 2016-2018, June 2016    74574
Name: perioddesc, dtype: int64
Colorado Short-Term Projections, 2016-2018, June 2016    74574
Name: perioddesc, dtype: int64


    CIM
    Colorado Short-Term Projections, 2016-2018, June 2016        372870
    Colorado Short-Term Projections, 2020-2022, December 2019     86350

    0324 We Pulled
    Colorado Short-Term Projections, 2016-2018, June 2016    74574

    0424 We Pulled
    Colorado Short-Term Projections, 2016-2018, June 2016    74574

    0324 Actual
    69	Colorado Short-Term Projections, 2016-2018, June 2016
    81	Colorado Short-Term Projections, 2022-2024, July 2023
    82	Colorado Short-Term Projections, 2023-2025, March 2024

    0424 Actual
    69	Colorado Short-Term Projections, 2016-2018, June 2016
    82	Colorado Short-Term Projections, 2023-2025, March 2024

In [11]:
dfs['lngproj.csv']['perioddesc'].value_counts()

Colorado Long-Term Projections, 2022-2032, July 2023    193594
Colorado Long-Term Projections, 2019-2029, June 2020    170496
Colorado Long-Term Projections, 2020-2030, July 2021     90055
Name: perioddesc, dtype: int64

In [16]:
print(dfs4['shproj.csv']['periodid'].value_counts())
print(dfs3['shproj.csv']['periodid'].value_counts())

69    74574
Name: periodid, dtype: int64
69    74574
Name: periodid, dtype: int64


In [8]:
ffs = ["Ces.csv","Income.csv","Industry.csv","Labforce.csv","OESWage.csv"]
for file in ffs:
    print(file)
    print(dfs[file]["periodyear"].value_counts().sort_index())
    print("---------------")

Ces.csv
1990    4431
1991    4431
1992    4431
1993    4431
1994    4431
1995    4431
1996    4431
1997    4431
1998    4431
1999    4431
2000    4457
2001    4457
2002    4470
2003    4535
2004    4561
2005    4561
2006    4561
2007    4561
2008    4561
2009    4561
2010    4561
2011    4561
2012    4561
2013    4561
2014    4561
2015    4561
2016    4561
2017    4561
2018    4561
2019    4535
2020    4535
2021    4535
2022    4535
2023    4431
2024     991
Name: periodyear, dtype: int64
---------------
Income.csv
1929      4
1930      4
1931      4
1932      4
1933      4
       ... 
2015    201
2016    197
2017    132
2020     65
2021     65
Name: periodyear, Length: 91, dtype: int64
---------------
Industry.csv
2000     60694
2001     63377
2002     64410
2003     64763
2004     64815
2005     61845
2006     64695
2007    106092
2008     53939
2009     53976
2010     52878
2011     51565
2012     51574
2013     51512
2014     51422
2015     51274
2016     51713
2017     51786
2018 

In [11]:
dfs["lngproj.csv"].head()

,stateabbrv,statename,stfips,areatype,areatyname,area,areaname,periodid,perioddesc,periodtype,matincodty,matincode,matintitle,matoccodty,matoccode,matocctitl,estemp,projemp,pctestind,pctprojind,nchg,aopeng,aopenr,aopent,suppress,pctprojocc,pctestocc,growrate,pchg
0,CO,Colorado,8,75,Balance of State,80001,Eastern and Southern Colorado,29,"Colorado Long-Term Projections, 2019-2029, Jun...",1,2,541000,"Professional, Scientific, and Technical Services",2,292000,Health Technologists and Technicians,76,86,6.95,7.60,10,NaN,NaN,NaN,0,4.32,4.01,1.2438,13.1579
1,CO,Colorado,8,21,Metropolitan Statistical Area,14500,Boulder-Longmont MSA,29,"Colorado Long-Term Projections, 2019-2029, Jun...",1,2,0,Total All Industries,2,191029,"Biological Scientists, All Other",213,265,0.10,0.12,52,NaN,NaN,NaN,0,100.00,100.00,2.2084,24.4131
2,CO,Colorado,8,21,Metropolitan Statistical Area,39380,Pueblo MSA,29,"Colorado Long-Term Projections, 2019-2029, Jun...",1,2,620000,Health Care and Social Assistance,2,110000,Management Occupations,368,422,2.70,2.74,54,NaN,NaN,NaN,0,14.62,13.77,1.3786,14.6739
3,CO,Colorado,8,1,State,0,Colorado,29,"Colorado Long-Term Projections, 2019-2029, Jun...",1,2,560000,Administrative and Support and Waste Managemen...,2,119000,Other Management Occupations,617,717,0.38,0.40,100,NaN,NaN,NaN,0,0.82,0.79,1.5134,16.2075
4,CO,Colorado,8,21,Metropolitan Statistical Area,19740,Denver - Aurora MSA,29,"Colorado Long-Term Projections, 2019-2029, Jun...",1,2,102000,Services Providing,2,193000,Social Scientists and Related Workers,2913,3395,0.21,0.23,482,NaN,NaN,NaN,0,79.81,79.85,1.5430,16.5465


In [16]:
display(dfs["lngproj.csv"]["perioddesc"].value_counts())

Colorado Long-Term Projections, 2022-2032, July 2023    193594
Colorado Long-Term Projections, 2019-2029, June 2020    170496
Colorado Long-Term Projections, 2020-2030, July 2021     90055
Name: perioddesc, dtype: int64

In [6]:
dfs["shproj.csv"].head()

,stateabbrv,statename,stfips,areatype,areatyname,areaname,periodid,perioddesc,periodtype,matincodty,matintitle,matoccodty,matocctitl,estemp,projemp,nchg,aopeng,aopenr,aopent,suppress,area,matincode,matoccode,pctestind,pctprojind,pchg,growrate,pctprojocc,pctestocc
0,CO,Colorado,8,21,Metropolitan Statistical Area,Colorado Springs MSA,76,"Colorado Short-Term Projections, 2020-2022, De...",1,2,Services Providing,2,Optometrists,98,99,1,NaN,NaN,NaN,0,17820,102000,291041,0.04,0.04,1.0204,0.5089,89.19,89.09
1,CO,Colorado,8,21,Metropolitan Statistical Area,Denver - Aurora MSA,76,"Colorado Short-Term Projections, 2020-2022, De...",1,2,"Self Employed and Unpaid Family Workers, All Jobs",2,"Shipping, Receiving, and Traffic Clerks",7,7,0,NaN,NaN,NaN,0,19740,67,435071,0.01,0.01,0.0000,0.0000,0.13,0.13
2,CO,Colorado,8,1,State,Colorado,76,"Colorado Short-Term Projections, 2020-2022, De...",1,2,Printing and Related Support Activities,2,"Installation, Maintenance, and Repair Occupations",43,40,-3,NaN,NaN,NaN,0,0,323000,490000,0.88,0.87,-6.9767,-3.5514,0.04,0.04
3,CO,Colorado,8,21,Metropolitan Statistical Area,Fort Collins-Loveland MSA,76,"Colorado Short-Term Projections, 2020-2022, De...",1,2,Finance and Insurance,2,Office and Administrative Support Occupations,1382,1381,-1,NaN,NaN,NaN,0,22660,520000,430000,37.73,37.28,-0.0724,-0.0362,6.09,6.06
4,CO,Colorado,8,75,Balance of State,Eastern and Southern Colorado,76,"Colorado Short-Term Projections, 2020-2022, De...",1,2,Government,2,Financial Specialists,194,197,3,NaN,NaN,NaN,0,80001,900000,132000,1.85,1.87,1.5464,0.7702,19.54,19.44


In [6]:
display(dfs["shproj.csv"]["perioddesc"].value_counts())
display(dfs["lngproj.csv"]["perioddesc"].value_counts())

Colorado Short-Term Projections, 2016-2018, June 2016        372870
Colorado Short-Term Projections, 2020-2022, December 2019     86350
Name: perioddesc, dtype: int64

Colorado Long-Term Projections, 2022-2032, July 2023    193594
Colorado Long-Term Projections, 2019-2029, June 2020    170496
Colorado Long-Term Projections, 2020-2030, July 2021     90055
Name: perioddesc, dtype: int64

In [79]:
for file,df in dfs.items():
    print(file,df.shape,dfs4[file].shape,dfs4[file].shape[0]-df.shape[0])
    # print(df.columns)
    # print(dfs4[file].columns)

Ces.csv (151495, 30) (154206, 30) 2711
Income.csv (8892, 19) (9022, 19) 130
Industry.csv (1346348, 29) (1372736, 29) 26388
Labforce.csv (34959, 18) (35598, 18) 639
OESWage.csv (284061, 40) (284676, 40) 615
lngproj.csv (260551, 29) (96797, 29) -163754
shproj.csv (310072, 29) (74574, 29) -235498


Ces.csv (151495, 30) (154206, 30) 2711   Replace
Income.csv (8892, 19) (9022, 19) 130     Replace
Industry.csv (1346348, 29) (1372736, 29) 26388  Replace
Labforce.csv (34959, 18) (35598, 18) 639  Replace
OESWage.csv (284061, 40) (284676, 40) 615 Replace
lngproj.csv (260551, 29) (96797, 29) -163754  Upsert
shproj.csv (310072, 29) (74574, 29) -235498  Upsert

In [8]:
dfs4.keys()

dict_keys(['Ces.csv', 'Income.csv', 'Industry.csv', 'Labforce.csv', 'OESWage.csv', 'lngproj.csv', 'shproj.csv'])

## Check Period of Record

In [11]:
dfs['Ces.csv'].columns

Index(['stateabbrv', 'statename', 'stfips', 'areatyname', 'areaname', 'area',
       'periodyear', 'periodtype', 'pertypdesc', 'period', 'seriescode',
       'seriesttls', 'seriesdesc', 'adjusted', 'benchmark', 'prelim', 'empces',
       'empprodwrk', 'empfemale', 'hours', 'earnings', 'hourearn',
       'supprecord', 'supphe', 'supppw', 'suppfem', 'hoursallwrkr',
       'earningsallwrkr', 'hourearnallwrkr', 'suppheallwrkr'],
      dtype='object')

In [15]:
print(dfs['Ces.csv']['pertypdesc'].value_counts())
print(dfs['Ces.csv']['periodtype'].value_counts())

Monthly    143404
Annual      10802
Name: pertypdesc, dtype: int64
3    143404
1     10802
Name: periodtype, dtype: int64


In [14]:
dfs['Ces.csv']['period'].value_counts()

2     12237
1     12237
12    11893
11    11893
10    11893
9     11893
8     11893
7     11893
6     11893
5     11893
4     11893
3     11893
0     10802
Name: period, dtype: int64

In [31]:
ffs = ["Ces.csv","Income.csv","Industry.csv","Labforce.csv","OESWage.csv"]
for file in ffs:
    print(file)
    print(dfs[file]['periodtype'].value_counts())
    print(dfs[file]['pertypdesc'].value_counts())
    
    

Ces.csv
3    143404
1     10802
Name: periodtype, dtype: int64
Monthly    143404
Annual      10802
Name: pertypdesc, dtype: int64
Income.csv
1    9022
Name: periodtype, dtype: int64
Annual    9022
Name: pertypdesc, dtype: int64
Industry.csv
2    1087830
1     258518
Name: periodtype, dtype: int64
Quarterly    1087830
Annual        258518
Name: pertypdesc, dtype: int64
Labforce.csv
3    33136
1     2462
Name: periodtype, dtype: int64
Monthly    33136
Annual      2462
Name: pertypdesc, dtype: int64
OESWage.csv
1    284061
Name: periodtype, dtype: int64
Annual    284061
Name: pertypdesc, dtype: int64


In [ ]:
### Labforce.csv -   [Unemployment Estimates in Colorado](https://data.colorado.gov/a/b/4e3w-qire) - "labforce"
### Industry.csv -   [Employee Counts by Industry in Colorado](https://data.colorado.gov/a/b/cjkq-q9ih) - "industry"
### Ces.csv -     [Hours Worked by Employees in Colorado](https://data.colorado.gov/a/b/pt2g-89wc) - "ces"
### Income.csv -  [Personal Income in Colorado](https://data.colorado.gov/a/b/2cpa-vbur) - "income"
### OESWage.csv - [Employment Wages in Colorado](https://data.colorado.gov/a/b/busm-qa5b) - "oes"

### lngproj.csv - [Long-Term Employment Projections in Colorado](https://data.colorado.gov/a/b/gyeb-jc69) - "lng proj, lmi"
### shproj.csv -  [Short-Term Employment Projections in Colorado](https://data.colorado.gov/a/b/u2t6-bfhr) - "shrt proj, lmi"


In [8]:
ffs = ["Ces.csv","Income.csv","Industry.csv","Labforce.csv","OESWage.csv"]
for file in ffs:
    invtYrMo={}
    print(file)
    print("Year  Jan  Feb  Mar  Apr  May  Jun  Jul  Aug  Sep  Oct  Nov  Dec")
    for index,row in dfs[file].loc[dfs[file]['periodtype']==3].iterrows():
        year=row['periodyear']
        month=row['period']

        if year not in invtYrMo:
            invtYrMo[year]={}
        if month not in invtYrMo[year]:
            invtYrMo[year][month]=0
        invtYrMo[year][month]+=1
        
    for year,invt in sorted(invtYrMo.items()):
        print(f"{year:4d}",end="")
        for mo in range(1,13):
            if mo in invt:
               cnt=invt[mo]
            else:
               cnt=0
            print(f"{cnt:5d}",end="")
        print()
    print()
    print()
        
          
        

Ces.csv
Year  Jan  Feb  Mar  Apr  May  Jun  Jul  Aug  Sep  Oct  Nov  Dec
1990  344  344  344  344  344  344  344  344  344  344  344  344
1991  344  344  344  344  344  344  344  344  344  344  344  344
1992  344  344  344  344  344  344  344  344  344  344  344  344
1993  344  344  344  344  344  344  344  344  344  344  344  344
1994  344  344  344  344  344  344  344  344  344  344  344  344
1995  344  344  344  344  344  344  344  344  344  344  344  344
1996  344  344  344  344  344  344  344  344  344  344  344  344
1997  344  344  344  344  344  344  344  344  344  344  344  344
1998  344  344  344  344  344  344  344  344  344  344  344  344
1999  344  344  344  344  344  344  344  344  344  344  344  344
2000  346  346  346  346  346  346  346  346  346  346  346  346
2001  346  346  346  346  346  346  346  346  346  346  346  346
2002  347  347  347  347  347  347  347  347  347  347  347  347
2003  352  352  352  352  352  352  352  352  352  352  352  352
2004  354  354  3

In [33]:
ffs = ["Ces.csv","Income.csv","Industry.csv","Labforce.csv","OESWage.csv"]
for file in ffs:
    invtYr={}
    print(file)
    for index,row in dfs[file].loc[dfs[file]['periodtype']==1].iterrows():
        year=row['periodyear']
        month=row['period']

        if year not in invtYr:
            invtYr[year]=0
        invtYr[year]+=1
      
    for year,cnt in sorted(invtYr.items()):
        print(f"{year:4d}  {cnt:5d}")

Ces.csv
1990    303
1991    303
1992    303
1993    303
1994    303
1995    303
1996    303
1997    303
1998    303
1999    303
2000    305
2001    305
2002    306
2003    311
2004    313
2005    313
2006    313
2007    313
2008    313
2009    313
2010    313
2011    313
2012    313
2013    313
2014    313
2015    313
2016    313
2017    313
2018    313
2019    311
2020    311
2021    311
2022    311
2023    303
2024    303
Income.csv
1929      4
1930      4
1931      4
1932      4
1933      4
1934      4
1935      4
1936      4
1937      4
1938      4
1939      4
1940      4
1941      4
1942      4
1943      4
1944      4
1945      4
1946      4
1947      4
1948      4
1949      4
1950      4
1951      4
1952      4
1953      4
1954      4
1955      4
1956      4
1957      4
1958      4
1959      4
1960      4
1961      4
1962      4
1963      4
1964      4
1965      4
1966      4
1967      4
1968      4
1969    146
1970    146
1971    146
1972    146
1973    146
1974    146
1975    1

In [23]:
for year,invt in sorted(invtYrMo.items()):
    print(f"{year:4d}",end="")
    for mo in range(1,13):
        if mo in invt:
           cnt=invt[mo]
        else:
           cnt=0
        print(f"{cnt:5d}",end="")
    print()
          
        

1990  344  344  344  344  344  344  344  344  344  344  344  344
1991  344  344  344  344  344  344  344  344  344  344  344  344
1992  344  344  344  344  344  344  344  344  344  344  344  344
1993  344  344  344  344  344  344  344  344  344  344  344  344
1994  344  344  344  344  344  344  344  344  344  344  344  344
1995  344  344  344  344  344  344  344  344  344  344  344  344
1996  344  344  344  344  344  344  344  344  344  344  344  344
1997  344  344  344  344  344  344  344  344  344  344  344  344
1998  344  344  344  344  344  344  344  344  344  344  344  344
1999  344  344  344  344  344  344  344  344  344  344  344  344
2000  346  346  346  346  346  346  346  346  346  346  346  346
2001  346  346  346  346  346  346  346  346  346  346  346  346
2002  347  347  347  347  347  347  347  347  347  347  347  347
2003  352  352  352  352  352  352  352  352  352  352  352  352
2004  354  354  354  354  354  354  354  354  354  354  354  354
2005  354  354  354  354 

## Short and Long Term Projections

In [19]:
print(dfs4['lngproj.csv']['perioddesc'].value_counts())
print(dfs3['lngproj.csv']['perioddesc'].value_counts())
print(dfs['lngproj.csv']['perioddesc'].value_counts())

Colorado Long-Term Projections, 2022-2032, July 2023    96797
Name: perioddesc, dtype: int64
Colorado Long-Term Projections, 2022-2032, July 2023    96797
Name: perioddesc, dtype: int64
Colorado Long-Term Projections, 2022-2032, July 2023    193594
Colorado Long-Term Projections, 2019-2029, June 2020    170496
Colorado Long-Term Projections, 2020-2030, July 2021     90055
Name: perioddesc, dtype: int64


### Long Term


CIM
Colorado Long-Term Projections, 2019-2029, June 2020    170496
Colorado Long-Term Projections, 2020-2030, July 2021     90055

March From CDLE
32	Colorado Long-Term Projections, 2022-2032, July 2023

April From CDLE
32	Colorado Long-Term Projections, 2022-2032, July 2023

Labels
Colorado Long-Term Projections, 2015-2025, August 2016
Colorado Long-Term Projections, 2016-2026, August 2017
Colorado Long-Term Projections, 2017-2027, June 2018
Colorado Long-Term Projections, 2021-2031, August 2022
Colorado Long-Term Projections, 2022-2032, July 2023



### Short Term


CIM
Colorado Short-Term Projections, 2016-2018, June 2016        372870
Colorado Short-Term Projections, 2020-2022, December 2019     86350

March CDLE Update
69	Colorado Short-Term Projections, 2016-2018, June 2016
81	Colorado Short-Term Projections, 2022-2024, July 2023
82	Colorado Short-Term Projections, 2023-2025, March 202404 

April CDLE Update
69	Colorado Short-Term Projections, 2016-2018, June 2016
82	Colorado Short-Term Projections, 2023-2025, March 2024

Labels
Colorado Short-Term Projections, 2016-2018, June 2016
Colorado Short-Term Projections, 2017-2019, Dec, 2016
Colorado Short-Term Projections, 2017-2019, August 2017
Colorado Short-Term Projections, 2019-2021, December 2018
Colorado Short-Term Projections, 2020-2022, December 2019
Colorado Short-Term Projections, 2022-2024, February 2023
Colorado Short-Term Projections, 2022-2024, July 2023
Colorado Short-Term Projections, 2023-2025, March 2024

In [12]:

dfs4['shproj.csv'].columns

Index(['Changed database context to 'GoCode_0424'.'], dtype='object')

In [18]:
print(dfs4['shproj.csv']['perioddesc'].value_counts())
print(dfs3['shproj.csv']['perioddesc'].value_counts())
print(dfs['shproj.csv']['perioddesc'].value_counts())

Colorado Short-Term Projections, 2023-2025, March 2024    91198
Colorado Short-Term Projections, 2016-2018, June 2016     74574
Name: perioddesc, dtype: int64
Colorado Short-Term Projections, 2022-2024, July 2023     96584
Colorado Short-Term Projections, 2023-2025, March 2024    91198
Colorado Short-Term Projections, 2016-2018, June 2016     74574
Name: perioddesc, dtype: int64
Colorado Short-Term Projections, 2016-2018, June 2016        372870
Colorado Short-Term Projections, 2020-2022, December 2019     86350
Name: perioddesc, dtype: int64


In [35]:
print(dfs4['shproj.csv']['periodid'].value_counts())
print(dfs3['shproj.csv']['periodid'].value_counts())
print(dfs['shproj.csv']['periodid'].value_counts())

69    74574
Name: periodid, dtype: int64
69    74574
Name: periodid, dtype: int64
69    372870
76     86350
Name: periodid, dtype: int64


In [13]:
dfs4['lngproj.csv'].compare(dfs3['lngproj.csv'])

areatype       areatyname           area                      areaname  \
          self other       self other     self    other                 self   
0          NaN   NaN        NaN   NaN  19740.0  17820.0  Denver - Aurora MSA   
1          NaN   NaN        NaN   NaN  19740.0  17820.0  Denver - Aurora MSA   
2          NaN   NaN        NaN   NaN  19740.0  17820.0  Denver - Aurora MSA   
3          NaN   NaN        NaN   NaN  19740.0  17820.0  Denver - Aurora MSA   
4          NaN   NaN        NaN   NaN  19740.0  17820.0  Denver - Aurora MSA   
...        ...   ...        ...   ...      ...      ...                  ...   
96792      NaN   NaN        NaN   NaN  80002.0  80003.0   Southwest Colorado   
96793      NaN   NaN        NaN   NaN  80002.0  80003.0   Southwest Colorado   
96794      NaN   NaN        NaN   NaN  80002.0  80003.0   Southwest Colorado   
96795      NaN   NaN        NaN   NaN  80002.0  80003.0   Southwest Colorado   
96796      NaN   NaN        NaN   NaN  80002.0  80003.0   Southwest Colorado   

                            matincode        ... pctprojind       pctprojocc  \
                      other      self other  ...       self other       self   
0      Colorado Springs MSA     670.0  67.0  ...       1.05  0.09       2.91   
1      Colorado Springs MSA     670.0  67.0  ...       1.76  0.02       4.42   
2      Colorado Springs MSA     670.0  67.0  ...       0.01  0.00       2.59   
3      Colorado Springs MSA     670.0  67.0  ...       0.97  0.00      12.91   
4      Colorado Springs MSA     670.0  67.0  ...       0.41  0.01       5.93   
...                     ...       ...   ...  ...        ...   ...        ...   
96792    Northwest Colorado       NaN   NaN  ...       1.35  3.43      79.67   
96793    Northwest Colorado       NaN   NaN  ...       0.82  0.66       2.51   
96794    Northwest Colorado       NaN   NaN  ...       0.20  2.41       2.14   
96795    Northwest Colorado       NaN   NaN  ...       0.39  0.46      18.34   
96796    Northwest Colorado       NaN   NaN  ...       0.18  0.37      76.00   

               nchg           pchg           growrate          
       other   self other     self     other     self   other  
0       1.25 -154.0   4.0 -10.4265   22.2222  -1.0951  2.0270  
1       1.68   70.0   0.0   3.2543    0.0000   0.3208  0.0000  
2       2.70    NaN   NaN      NaN       NaN      NaN     NaN  
3       0.49   -7.0   0.0  -0.5705    0.0000  -0.0572  0.0000  
4       0.98   35.0   1.0   7.2464  100.0000   0.7020  7.1773  
...      ...    ...   ...      ...       ...      ...     ...  
96792   7.64   22.0  63.0  17.8862   20.7237   1.6591  1.9012  
96793   4.71    NaN   NaN  15.7895   20.3390   1.4768  1.8687  
96794  86.87    3.0  44.0  16.6667   20.5607   1.5534  1.8874  
96795   1.45    6.0   9.0  16.6667   22.5000   1.5534  2.0501  
96796  58.82    3.0   7.0  18.7500   21.2121   1.7334  1.9423  

[96797 rows x 34 columns]

In [16]:
dfs4["Income.csv"].loc[(dfs4["Income.csv"]['periodyear'] == 2024) & (dfs4["Income.csv"]['periodtype'] != 0),'periodtype'].value_counts()

Series([], Name: periodtype, dtype: int64)

In [24]:
dfs4["Ces.csv"].columns

Index(['stateabbrv', 'statename', 'stfips', 'areatyname', 'areaname', 'area',
       'periodyear', 'periodtype', 'pertypdesc', 'period', 'seriescode',
       'seriesttls', 'seriesdesc', 'adjusted', 'benchmark', 'prelim', 'empces',
       'empprodwrk', 'empfemale', 'hours', 'earnings', 'hourearn',
       'supprecord', 'supphe', 'supppw', 'suppfem', 'hoursallwrkr',
       'earningsallwrkr', 'hourearnallwrkr', 'suppheallwrkr'],
      dtype='object')

In [25]:
dfs4["Ces.csv"].loc[(dfs4["Ces.csv"]['periodyear'] == 2024),'period'].value_counts()

2    344
1    344
0    303
Name: period, dtype: int64

In [27]:
dfs3["Ces.csv"].loc[(dfs3["Ces.csv"]['periodyear'] == 2023),'period'].value_counts()

11    344
7     344
3     344
10    344
6     344
2     344
9     344
5     344
1     344
12    344
8     344
4     344
0     303
Name: period, dtype: int64

In [21]:
dfs3["Income.csv"]['periodyear'].value_counts().sort_index()

1929      4
1930      4
1931      4
1932      4
1933      4
       ... 
2015    201
2016    197
2017    132
2020     65
2021     65
Name: periodyear, Length: 91, dtype: int64

In [29]:
dfs["Ces.csv"].loc[(dfs["Ces.csv"]['periodyear'] == 2023),'period'].value_counts()

7    344
5    344
3    344
1    344
6    344
4    344
2    344
0    303
Name: period, dtype: int64

In [12]:
dfLng = pd.read_csv("/home/joe/bic_etl/cdle/cdle_LaborAndEmployment/data_transformed/lngproj.csv")
dfSh = pd.read_csv("/home/joe/bic_etl/cdle/cdle_LaborAndEmployment/data_transformed/shproj.csv")

In [13]:
dfLng.head()

,stateabbrv,statename,stfips,areatype,areatyname,area,areaname,periodid,perioddesc,periodtype,matincodty,matincode,matintitle,matoccodty,matoccode,matocctitl,estemp,projemp,pctestind,pctestocc,pctprojind,pctprojocc,nchg,pchg,growrate,aopeng,aopenr,aopent,suppress
0,CO,Colorado,8,21,Metropolitan Statistical Area,19740,Denver - Aurora MSA,32,"Colorado Long-Term Projections, 2022-2032, Jul...",1,2,0,Total All Industries,2,0,"Total, All Occupations",1716301,1967762,100.0,100.00,100.0,100.0,251461,14.6513,1.3766,NaN,NaN,NaN,0
1,CO,Colorado,8,21,Metropolitan Statistical Area,19740,Denver - Aurora MSA,32,"Colorado Long-Term Projections, 2022-2032, Jul...",1,2,67,"Self Employed and Unpaid Family Workers, All Jobs",2,0,"Total, All Occupations",116133,125966,100.0,6.77,100.0,6.4,9833,8.4670,0.8161,NaN,NaN,NaN,0
2,CO,Colorado,8,21,Metropolitan Statistical Area,19740,Denver - Aurora MSA,32,"Colorado Long-Term Projections, 2022-2032, Jul...",1,2,670,"Self Employed and Unpaid Family Workers, All Jobs",2,0,"Total, All Occupations",116133,125966,100.0,6.77,100.0,6.4,9833,8.4670,0.8161,NaN,NaN,NaN,0
3,CO,Colorado,8,21,Metropolitan Statistical Area,19740,Denver - Aurora MSA,32,"Colorado Long-Term Projections, 2022-2032, Jul...",1,2,671,"Total Self Employed and Unpaid Family Workers,...",2,0,"Total, All Occupations",116133,125966,100.0,6.77,100.0,6.4,9833,8.4670,0.8161,NaN,NaN,NaN,0
4,CO,Colorado,8,21,Metropolitan Statistical Area,19740,Denver - Aurora MSA,32,"Colorado Long-Term Projections, 2022-2032, Jul...",1,2,6010,"Self Employed Workers, All Jobs",2,0,"Total, All Occupations",116133,125966,100.0,6.77,100.0,6.4,9833,8.4670,0.8161,NaN,NaN,NaN,0


In [14]:
dfLng.shape

(96797, 29)

In [15]:
dfLng['perioddesc'].value_counts()

Colorado Long-Term Projections, 2022-2032, July 2023    96797
Name: perioddesc, dtype: int64

In [16]:
dfSh['perioddesc'].value_counts()

Colorado Short-Term Projections, 2023-2025, March 2024    91198
Name: perioddesc, dtype: int64

In [2]:
dfLngC = pd.read_csv("https://data.colorado.gov/resource/gyeb-jc69.csv?$limit=10000000")
dfShC = pd.read_csv("https://data.colorado.gov/resource/u2t6-bfhr.csv?$limit=10000000")

In [3]:
dfShC['perioddesc'].value_counts()

Colorado Short-Term Projections, 2023-2025, March 2024    91198
Name: perioddesc, dtype: int64

In [4]:
dfLngC['perioddesc'].value_counts()

Colorado Long-Term Projections, 2022-2032, July 2023    96797
Name: perioddesc, dtype: int64

In [36]:
print(dfLngC.shape)
print(dfShC.shape)

(96797, 29)
(91198, 29)


## School Programs

In [2]:
dfSchoolC = pd.read_csv("https://data.colorado.gov/resource/jnj7-fw37.csv?$limit=10000000")

In [3]:
dfSchoolC.shape

(10304, 13)

In [47]:
dfSchoolC.head()

,statename,city,areaname,areadesc,instname1,progtitle,progdesc,instowndes,insttydesc,latitude,longitude,url,zip
0,Colorado,NaN,Douglas County,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,Accounting and Business/Management.,NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN
1,Colorado,NaN,Douglas County,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,Accounting.,NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN
2,Colorado,NaN,Douglas County,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,Behavioral Sciences.,NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN
3,Colorado,NaN,Douglas County,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,"Business Administration and Management, General.",NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN
4,Colorado,NaN,Douglas County,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,"Computer and Information Sciences,",NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN


In [4]:
dfSchool = pd.read_csv("/home/joe/bic_etl/cdle/cdle_education/data_transformed/school_programs.csv")


In [5]:
dfSchool.shape

(10304, 13)

In [6]:
dfSchool.columns

Index(['statename', 'city', 'areaname', 'areadesc', 'instname1', 'progtitle',
       'progdesc', 'instowndes', 'insttydesc', 'latitude', 'longitude', 'url',
       'zip'],
      dtype='object')

In [13]:
dfSchool['longitude'].nunique()

120

In [9]:
dfSchool['instname1'].nunique()

NameError: name 'dfSchoolC' is not defined

In [72]:
dct = dict(zip(dfSchool["instname1"],dfSchool["progtitle"]))
dctC = dict(zip(dfSchoolC["instname1"],dfSchoolC["progtitle"]))

In [73]:
for inst,prg in dct.items():
    prgC = dctC[inst]
    if prg != prgC:
        print(inst,prg,prgC)

Intellitec College-Colorado Springs Physical Fitness Technician. Physical Fitness Technician. (NEW)
Pikes Peak Community College Wildland/Forest Firefighting and Investigation. Wildland/Forest Firefighting and Investigation. (NEW)
Heritage College-Denver Physical Fitness Technician. Physical Fitness Technician. (NEW)
Intellitec College-Grand Junction Physical Fitness Technician. Physical Fitness Technician. (NEW)
Aims Community College Wildland/Forest Firefighting and Investigation. Wildland/Forest Firefighting and Investigation. (NEW)


In [75]:
for instC,prgC in dctC.items():
    prg = dct[instC]
    if prg != prgC:
        print(instC)
        print(prg)        
        print(prgC)   
        print("---------")

Intellitec College-Colorado Springs
Physical Fitness Technician.
Physical Fitness Technician. (NEW)
---------
Pikes Peak Community College
Wildland/Forest Firefighting and Investigation.
Wildland/Forest Firefighting and Investigation. (NEW)
---------
Heritage College-Denver
Physical Fitness Technician.
Physical Fitness Technician. (NEW)
---------
Intellitec College-Grand Junction
Physical Fitness Technician.
Physical Fitness Technician. (NEW)
---------
Aims Community College
Wildland/Forest Firefighting and Investigation.
Wildland/Forest Firefighting and Investigation. (NEW)
---------


In [ ]:
fin = open("/home/joe/bic_etl/cdle/cdle_education/data_transformed/school_programs.csv")
hist={}
for line in fin:
    spl = line.split(",")
    nn=len(spl)
    if nn in hist:
        hist[nn]+=1
    else:
        hist[nn]=1
        
    if nn != 13:
        print(nn)
        print(line)
        print("--------")

In [77]:
hist

{13: 1153,
 18: 597,
 19: 813,
 22: 416,
 23: 372,
 14: 1689,
 15: 814,
 16: 786,
 20: 613,
 17: 635,
 25: 152,
 24: 193,
 26: 239,
 31: 88,
 28: 363,
 21: 327,
 27: 550,
 29: 148,
 30: 142,
 32: 89,
 33: 74,
 37: 8,
 34: 22,
 35: 13,
 36: 4,
 38: 5}

In [79]:
dfSchool.shape

(10304, 13)

In [27]:
dfSchool.to_csv("school.csv",index=False)

In [29]:
dfSchoolNew = pd.read_csv("school.csv")

In [17]:
dfSchoolNew.shape

(10304, 13)

In [84]:
dfSchoolC['areadesc'].values

array(["Douglas County is the eighth most populous of the 64 counties of the state of Colorado, in the United States. The county is located midway between Colorado's two largest cities: Denver and Colorado Springs. The United States Census Bureau that the county population was 285,465 in 2010 census, a 62.4% increase since the 2000 census, making Douglas County one of the fastest growing counties in the United States. Douglas County is part of the Denver-Aurora Metropolitan Statistical Area and the Denver-Aurora-Boulder Combined Statistical Area. The county seat is Castle Rock. Douglas County has the highest median household income of any Colorado county or statistical equivalent. It is ranked eighth nationally in that category, and has the highest of any county or equivalent not in the northeastern US.",
       "Douglas County is the eighth most populous of the 64 counties of the state of Colorado, in the United States. The county is located midway between Colorado's two largest citie

In [7]:
histC={}
for col in dfSchoolC.columns:
    vals = dfSchoolC[col].values.tolist()
    for val in vals:
        if isinstance(val,str):
           if col not in histC:
                histC[col]={}
           for c in val:
               nn = ord(c)
               if nn in histC[col]:
                    histC[col][nn]+=1
               else:
                    histC[col][nn]=1
            

In [ ]:
for col,chrs in histC.items():
    for chr,cnt in sorted(chrs.items()):
        print(col,chr,cnt)
    print("------------------")

In [ ]:
vals = dfSchoolC['areadesc'].values.tolist()

for val in vals:
    hit=False
    if isinstance(val,str):
        for nn,chr in enumerate(val):
            if ord(chr) == 34:
                print(nn,chr,ord(chr))
                hit=True
        if hit:
            print(val)
        print("---")

In [19]:
dfSchool['progdesc'].value_counts()

A program that generally prepares individuals to plan, organize, direct, and control the functions and processes of a firm or organization.  Includes instruction in management theory, human resources management and behavior, accounting and other quantitat,Public Institution"                46
Any single instructional program in liberal arts and sciences, general studies and humanities not listed above.                                                                                                                                                                                    45
A program that generally prepares individuals to cut, trim, and style scalp, facial, and body hair; apply cosmetic preparations; perform manicures and pedicures; massage the head and extremities; and prepare for practice as licensed cosmetologists in spec,Private for profit institution"    44
A program that is a structured combination of the arts, biological and physical sciences, social sciences, and humanit

In [20]:
dfSchool.isna().sum()

statename        0
city            85
areaname         0
areadesc      1243
instname1        0
progtitle        0
progdesc      5542
instowndes       0
insttydesc       0
latitude         0
longitude        0
url              0
zip           3914
dtype: int64

In [21]:
dfSchool.shape

(10304, 13)

In [22]:
dfSchool.columns

Index(['statename', 'city', 'areaname', 'areadesc', 'instname1', 'progtitle',
       'progdesc', 'instowndes', 'insttydesc', 'latitude', 'longitude', 'url',
       'zip'],
      dtype='object')

In [ ]:
dfSchool['instname1'].value_counts()

In [ ]:
dfSchoolNew.loc[dfSchoolNew['instname1'] == "Western State Colorado University"]

In [30]:
dfSchool = pd.read_csv("/home/joe/bic_etl/cdle/cdle_education/data_transformed/school_programs.csv",encoding="latin")


In [31]:
dfSchool.loc[dfSchool['instname1'] == "Western State Colorado University"]

KeyError: 'instname1'

In [3]:
dfSchoolMan = pd.read_csv("/home/joe/bic_etl/cdle/cdle_education/data_transformed/school_programs_manual.csv")


In [4]:
dfSchoolMan.columns

Index(['statename', 'city', 'address', 'areaname', 'areadesc', 'instname1',
       'progtitle', 'progdesc', 'instowndes', 'insttydesc', 'latitude',
       'longitude', 'url', 'zip'],
      dtype='object')

In [ ]:
dfSchoolMan['zip'].value_counts()

In [8]:
dfSchoolMan.isna().sum()

statename        0
city            85
address         85
areaname         0
areadesc      1243
instname1        0
progtitle        0
progdesc      5542
instowndes       0
insttydesc       0
latitude         0
longitude        0
url              0
zip             85
dtype: int64

In [15]:
dfSchoolMan.groupby(["instname1","progtitle",'address']).count()

statename  \
instname1                               progtitle                                          address                        
Academy of Natural Therapy Inc          Massage Therapy/Therapeutic Massage.               625 8th Ave                2   
                                        Somatic Bodywork.                                  625 8th Ave                1   
Adams State University                  American Government and Politics (United States).  208 Edgemont Blvd          8   
                                        Biology/Biological Sciences, General.              208 Edgemont Blvd          8   
                                        Business Administration and Management, General.   208 Edgemont Blvd          8   
...                                                                                                                 ...   
Westwood College-Denver North           Web Page, Digital/Multimedia and Information Re... 7350 N Broadway            7   
Xenon International Academy-Denver      Aesthetician/Esthetician and Skin Care Specialist. 2231 S Peoria             10   
                                        Cosmetology/Cosmetologist, General.                2231 S Peoria             10   
                                        Nail Technician/Specialist and Manicurist.         2231 S Peoria              5   
Yeshiva Toras Chaim Talmudical Seminary Talmudic Studies.                                  1555 Stuart St             1   

                                                                                                              city  \
instname1                               progtitle                                          address                   
Academy of Natural Therapy Inc          Massage Therapy/Therapeutic Massage.               625 8th Ave           2   
                                        Somatic Bodywork.                                  625 8th Ave           1   
Adams State University                  American Government and Politics (United States).  208 Edgemont Blvd     8   
                                        Biology/Biological Sciences, General.              208 Edgemont Blvd     8   
                                        Business Administration and Management, General.   208 Edgemont Blvd     8   
...                                                                                                            ...   
Westwood College-Denver North           Web Page, Digital/Multimedia and Information Re... 7350 N Broadway       7   
Xenon International Academy-Denver      Aesthetician/Esthetician and Skin Care Specialist. 2231 S Peoria        10   
                                        Cosmetology/Cosmetologist, General.                2231 S Peoria        10   
                                        Nail Technician/Specialist and Manicurist.         2231 S Peoria         5   
Yeshiva Toras Chaim Talmudical Seminary Talmudic Studies.                                  1555 Stuart St        1   

                                                                                                              areaname  \
instname1                               progtitle                                          address                       
Academy of Natural Therapy Inc          Massage Therapy/Therapeutic Massage.               625 8th Ave               2   
                                        Somatic Bodywork.                                  625 8th Ave               1   
Adams State University                  American Government and Politics (United States).  208 Edgemont Blvd         8   
                                        Biology/Biological Sciences, General.              208 Edgemont Blvd         8   
                                        Business Administration and Management, General.   208 Edgemont Blvd         8   
...                                                                                                                ...   

In [23]:
dfSchoolMan.loc[(dfSchoolMan['instname1'] == "Adams State University")  &
                (dfSchoolMan['progtitle'] == "American Government and Politics (United States).")]

,statename,city,address,areaname,areadesc,instname1,progtitle,progdesc,instowndes,insttydesc,latitude,longitude,url,zip
85,Colorado,Alamosa,208 Edgemont Blvd,Alamosa County ...,Alamosa County is one of the 64 counties of th...,Adams State University,American Government and Politics (United States).,NaN,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
86,Colorado,Alamosa,208 Edgemont Blvd,Alamosa County ...,Alamosa County is one of the 64 counties of th...,Adams State University,American Government and Politics (United States).,A program that focuses on the systematic study...,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
143,Colorado,Alamosa,208 Edgemont Blvd,Co Rural Workforce Consortium,NaN,Adams State University,American Government and Politics (United States).,NaN,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
144,Colorado,Alamosa,208 Edgemont Blvd,Co Rural Workforce Consortium,NaN,Adams State University,American Government and Politics (United States).,A program that focuses on the systematic study...,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
201,Colorado,Alamosa,208 Edgemont Blvd,Edwards MCA ...,Eagle & Lake counties,Adams State University,American Government and Politics (United States).,NaN,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
202,Colorado,Alamosa,208 Edgemont Blvd,Edwards MCA ...,Eagle & Lake counties,Adams State University,American Government and Politics (United States).,A program that focuses on the systematic study...,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
259,Colorado,Alamosa,208 Edgemont Blvd,Planning Region 3 ...,,Adams State University,American Government and Politics (United States).,NaN,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
260,Colorado,Alamosa,208 Edgemont Blvd,Planning Region 3 ...,,Adams State University,American Government and Politics (United States).,A program that focuses on the systematic study...,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0


In [ ]:
for inst in dfSchoolMan['instname1'].unique():
    dftmp = dfSchoolMan.loc[dfSchoolMan['instname1'] == inst]
#    dftmp.sort_values(by=['progtitle','areaname'],inplace=True)
    for 
    

In [5]:
dfbad = dfSchoolMan.loc[dfSchoolMan['progdesc'].isna()]
dfgood = dfSchoolMan.loc[~dfSchoolMan['progdesc'].isna()]

In [6]:
tot=0
histbad = {}
dftmp = pd.DataFrame(columns=dfgood.columns)
dfdel = pd.DataFrame(columns=dfgood.columns)

for index,row in dfbad.iterrows():
    inst = row['instname1']
    prog = row['progtitle']
    area = row['areaname']
    url = row['url']
    ownd = row['instowndes']
    indes = row['insttydesc']
    if inst not in histbad:
        histbad[inst]={}
        histbad[inst]['count']=0
        histbad[inst]['bad']=0
        
    histbad[inst]['count']+=1
    nn = dfgood.loc[(dfgood['instname1'] == inst) & 
                    (dfgood['progtitle'] == prog) & 
                    (dfgood['areaname']==area) & 
                    (dfgood['url'] == url) & 
                    (dfgood['instowndes'] == ownd) & 
                    (dfgood['insttydesc'] == indes)                   
                   ].shape[0]
    if nn == 0:
        tot+=1
        histbad[inst]['bad']+=1
        dftmp = dftmp.append(row)
    elif nn == 1:
        dfdel=dfdel.append(row)
    else:
        print(inst,prog,area)
        
        
print(dfbad.shape[0],tot)
  

5542 1873


In [7]:
dfdel.shape

(3669, 14)

In [ ]:
for inst,res in histbad.items():
    print(inst,res['count'],res['bad'])

In [49]:
dfdel.head()

,statename,city,address,areaname,areadesc,instname1,progtitle,progdesc,instowndes,insttydesc,latitude,longitude,url,zip
85,Colorado,Alamosa,208 Edgemont Blvd,Alamosa County ...,Alamosa County is one of the 64 counties of th...,Adams State University,American Government and Politics (United States).,NaN,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
87,Colorado,Alamosa,208 Edgemont Blvd,Alamosa County ...,Alamosa County is one of the 64 counties of th...,Adams State University,"Biology/Biological Sciences, General.",NaN,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
89,Colorado,Alamosa,208 Edgemont Blvd,Alamosa County ...,Alamosa County is one of the 64 counties of th...,Adams State University,"Business Administration and Management, General.",NaN,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
91,Colorado,Alamosa,208 Edgemont Blvd,Alamosa County ...,Alamosa County is one of the 64 counties of th...,Adams State University,"Business Administration, Management and Operat...",NaN,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0
93,Colorado,Alamosa,208 Edgemont Blvd,Alamosa County ...,Alamosa County is one of the 64 counties of th...,Adams State University,"Business/Commerce, General.",NaN,Public Institution,Four-year Colleges and Universities,0.0,0.0,http://www.adams.edu,81101.0


In [35]:
4762+1873

6635

In [36]:
5542-1873

3669

In [8]:
dfgood = pd.concat([dfgood,dftmp])

In [13]:
for index,row in dfdel.iterrows():
    inst = row['instname1']
    prog = row['progtitle']
    area = row['areaname']
    url = row['url']
    ownd = row['instowndes']
    indes = row['insttydesc']
 
    histbad[inst]['count']+=1
    hld= dfgood.loc[(dfgood['instname1'] == inst) & 
                    (dfgood['progtitle'] == prog) & 
                    (dfgood['areaname']==area) & 
                    (dfgood['url'] == url) & 
                    (dfgood['instowndes'] == ownd) & 
                    (dfgood['insttydesc'] == indes)                   
                   ]
    nn=hld.shape[0]
    
    if nn == 1:
        mm = hld['progdesc'].isna().sum()
        if mm != 0:
            print("ROH ROH")
            print(hld.head())
            print("----")
    else:
        print("BAD BAD BAD",nn)
        print(hld.head())
        print("----")

In [14]:
longs = dfSchool['longitude'].values.tolist()

In [20]:
display(dfSchool.iloc[85])


statename                                              Colorado
city                                                    Alamosa
areaname      Alamosa County                                ...
areadesc      Alamosa County is one of the 64 counties of th...
instname1                                Adams State University
progtitle     American Government and Politics (United States).
progdesc                                                    NaN
instowndes                                   Public Institution
insttydesc                  Four-year Colleges and Universities
latitude                                                    0.0
longitude                                              0.000000
url                                        http://www.adams.edu
zip                                                     81101.0
Name: 85, dtype: object

In [18]:
dfSchool.iloc[86]

statename                                              Colorado
city                                                    Alamosa
areaname      Alamosa County                                ...
areadesc      Alamosa County is one of the 64 counties of th...
instname1                                Adams State University
progtitle     American Government and Politics (United States).
progdesc      A program that focuses on the systematic study...
instowndes                  Four-year Colleges and Universities
insttydesc                                             0.000000
latitude                                                    0.0
longitude                                  http://www.adams.edu
url                                                       81101
zip                                                         NaN
Name: 86, dtype: object

In [ ]:
for nn,lng in enumerate(longs):
    print(nn,lng)

In [22]:
file = "/home/joe/bic_etl/cdle/cdle_education/data_transformed/school_programs.csv"
fin = open(file)
lines = fin.readlines()


In [24]:
lines[86]

'"Colorado","Alamosa","Alamosa County                                              ","Alamosa County is one of the 64 counties of the state of Colorado of the United States. The county name is the Spanish language word for a ""grove of cottonwood trees."" The county population was 14,966 at U.S. Census 2000. The county seat is Alamosa.","Adams State University","American Government and Politics (United States).",NULL,"Public Institution","Four-year Colleges and Universities","0.000000","0.000000","http://www.adams.edu","81101"\n'

In [25]:
lines[87]

'"Colorado","Alamosa","Alamosa County                                              ","Alamosa County is one of the 64 counties of the state of Colorado of the United States. The county name is the Spanish language word for a ""grove of cottonwood trees."" The county population was 14,966 at U.S. Census 2000. The county seat is Alamosa.","Adams State University","American Government and Politics (United States).","A program that focuses on the systematic study of United States political institutions and behavior.  Includes instruction in American political theory, political parties and interest groups, state and local governments, Constitutional law, federalism and,"Public Institution","Four-year Colleges and Universities","0.000000","0.000000","http://www.adams.edu","81101"\n'

In [26]:
schoolNew = pd.read_csv("school_programs.csv")

In [27]:
schoolNew.loc[schoolNew['instname1'] == "Western State Colorado University"]


KeyError: 'instname1'

In [28]:
schoolNew.head()

,"""Colorado""",NULL,NULL.1,"""Douglas County ""","""Douglas County is the eighth most populous of the 64 counties of the state of Colorado, in the United States. The county is located midway between Colorado's two largest cities: Denver and Colorado Springs. The United States Census Bureau that the county population was 285,465 in 2010 census, a 62.4% increase since the 2000 census, making Douglas County one of the fastest growing counties in the United States. Douglas County is part of the Denver-Aurora Metropolitan Statistical Area and the Denver-Aurora-Boulder Combined Statistical Area. The county seat is Castle Rock. Douglas County has the highest median household income of any Colorado county or statistical equivalent. It is ranked eighth nationally in that category, and has the highest of any county or equivalent not in the northeastern US.""","""University of Phoenix-Colorado Campus""","""Accounting and Business/Management.""",NULL.2,"""Private for profit institution""","""Four-year Colleges and Universities""","""39.551643""","""-104.873421""","""http://""",NULL.3
0,"""Colorado""",NaN,NaN,"""Douglas County ...","""Douglas County is the eighth most populous of...","""University of Phoenix-Colorado Campus""","""Accounting.""",NaN,"""Private for profit institution""","""Four-year Colleges and Universities""","""39.551643""","""-104.873421""","""http://""",NaN
1,"""Colorado""",NaN,NaN,"""Douglas County ...","""Douglas County is the eighth most populous of...","""University of Phoenix-Colorado Campus""","""Behavioral Sciences.""",NaN,"""Private for profit institution""","""Four-year Colleges and Universities""","""39.551643""","""-104.873421""","""http://""",NaN
2,"""Colorado""",NaN,NaN,"""Douglas County ...","""Douglas County is the eighth most populous of...","""University of Phoenix-Colorado Campus""","""Business Administration and Management, Gener...",NaN,"""Private for profit institution""","""Four-year Colleges and Universities""","""39.551643""","""-104.873421""","""http://""",NaN
3,"""Colorado""",NaN,NaN,"""Douglas County ...","""Douglas County is the eighth most populous of...","""University of Phoenix-Colorado Campus""","""Computer and Information Sciences,""",NaN,"""Private for profit institution""","""Four-year Colleges and Universities""","""39.551643""","""-104.873421""","""http://""",NaN
4,"""Colorado""",NaN,NaN,"""Douglas County ...","""Douglas County is the eighth most populous of...","""University of Phoenix-Colorado Campus""","""Computer and Information Systems Security/Inf...",NaN,"""Private for profit institution""","""Four-year Colleges and Universities""","""39.551643""","""-104.873421""","""http://""",NaN


In [2]:
df = pd.read_csv("school_programs.csv")

In [3]:
df.head()

,statename,city,areaname,areadesc,instname1,progtitle,progdesc,instowndes,insttydesc,latitude,longitude,url,zip
0,Colorado,NaN,Douglas County ...,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,Accounting and Business/Management.,NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN
1,Colorado,NaN,Douglas County ...,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,Accounting.,NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN
2,Colorado,NaN,Douglas County ...,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,Behavioral Sciences.,NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN
3,Colorado,NaN,Douglas County ...,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,"Business Administration and Management, General.",NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN
4,Colorado,NaN,Douglas County ...,Douglas County is the eighth most populous of ...,University of Phoenix-Colorado Campus,"Computer and Information Sciences,",NaN,Private for profit institution,Four-year Colleges and Universities,39.551643,-104.873421,http://,NaN


In [4]:
df['longitude'].value_counts()

 0.000000      9714
-104.986386     196
-105.000000     154
-104.822841      66
-104.873421      60
-104.820000      30
-104.980000      28
-104.875949      20
-104.821001      13
-104.780941       7
-104.787967       5
-105.053321       4
-105.200000       3
-104.967732       1
-105.043282       1
-105.080647       1
-105.065051       1
Name: longitude, dtype: int64